roofit/roofitcore/inc/RangeCast.h

```diff
@@ -0,0 +1,111 @@
+// Author: Jonas Rembser, CERN, Jul 2021
+
+#ifndef roofit_roofitcore_RangeCast_h
+#define roofit_roofitcore_RangeCast_h
+
+namespace detail {
+
+template <bool isDynamic, typename T>
+struct TemplatedCaster {
+   template <typename U>
+   static T cast(U const &u);
+};
+
+template <typename T>
+struct TemplatedCaster<false, T> {
+   template <typename U>
+   static T cast(U const &u)
+   {
+      return static_cast<T>(u);
+   }
+};
+
+template <typename T>
+struct TemplatedCaster<true, T> {
+   template <typename U>
+   static T cast(U const &u)
+   {
+      return dynamic_cast<T>(u);
+   }
+};
+
+} // namespace detail
+
+template <typename T, typename Range_t, bool isDynamic>
+class RangeCaster {
+
+public:
+   using const_iterator_orig = typename Range_t::const_iterator;
+
+   class CastingIterator {
+
+   public:
+      CastingIterator(const_iterator_orig const &iter) : _iter{iter} {}
+
+      CastingIterator &operator++()
+      {
+         ++_iter;
+         return *this;
+      }
+      CastingIterator operator++(int)
+      {
+         CastingIterator tmp(*this);
+         operator++();
+         return tmp;
+      }
+      bool operator==(const CastingIterator &rhs) const { return _iter == rhs._iter; }
+      bool operator!=(const CastingIterator &rhs) const { return _iter != rhs._iter; }
+      T operator*() { return detail::TemplatedCaster<isDynamic, T>::cast(*_iter); }
```

Comment:
I assume that the compiler will do equally well as for `if constexpr` if you use this:
```c++
if (isDynamic)
  return dynamic_cast<T>(*_iter);
else
  return static_cast<T>(*_iter);
```
You can do similarly for `operator->`.

Reply:
Further, think about
```c++
else {
  assert(dynamic_cast<T>(*_iter));
  return static_cast<T>(*_iter);
}
```
This might save you from making a few not-so-easy-to-find mistakes.

Reply:
Good idea!

Reply:
Note that `static_cast` is free in terms of runtime, `dynamic_cast` (used e.g. in the `assert` above) might add a little overhead when RTTI is required which users of `static_cast` might not expect.

(I don't know if these adapters will be used in hot loops a lot, so might be not relevant, feel free to ignore)

Reply:
These will be used in hot loops. But the RooFit philosophy so far was to ignore the performance penalty intoduced by `assert` because it is ignored in the release builds.

I see now that this is an unexpected behavior if `static_range_cast` is used by people other than the RooFit developers, so actually I think it's better to not do the assert here.

Reply:
Your decision. I'm just saying that those asserts helped trace down a ton of bugs, and it's really as easy as running the thing that somebody sends you in your debug build.

Reply:
sorry I take back my comment about performance, `assert`s are totally fine in performance-critical code!